### PR TITLE
NEW: Properly register the plugin and generated methods with the framework

### PR DESCRIPTION
**Pull request by Oddant1**
The generated methods are now registered along with the plugin and are accessible through the CLI

### CODE REVIEW

**Comment by Oddant1** [comment]
Looks like the .gitignore has decided to let my .vscode folder in even though it says to ingore them

**Comment by Oddant1** [comment]
At this point, the tests don't assert anything other than the implicit assertion that the usage examples have been properly created.

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,29 @@
+BSD 3-Clause License
+
+Copyright (c) 2016-2020, QIIME 2 development team.
```
Comment (thermokarst) [inline LICENSE]: Drive-by-comment: the copyright should just be for 2020, since that is the year this plugin was created in.

**Comment by Oddant1** [comment]
https://github.com/qiime2/q2-mystery-stew/pull/3/commits/d9a10d8f6c4ad62689c26af69a7f1d4800062dfa Most of the `# TODO: Remove` stuff in this commit has to do with the fact that the metadata args aren't regexing properly, and I don't know why yet.

**Comment by Oddant1** [comment]
Do not attempt to run the test on https://github.com/qiime2/q2-mystery-stew/pull/3/commits/fe23ca4df031cb01d3d2a919d857e38f9d398937. It will probably take literal hours to finish. This is just a proof of concept for some attempts I'm making at narrowing down the tests (yes this commit does actually create less usage examples than some of the previous ones, I know this because it can at least finish creating the examples while some of the previous commits locked my computer up and I had to reboot while attempting to create usage examples).

**Comment by Oddant1** [comment]
As of https://github.com/qiime2/q2-mystery-stew/pull/3/commits/e8393df4fe59fce150d61801e2606eeeee1f8dbe Lists and Sets should work. We are generating some 127944. In theory all of these aside from those involving collections shoult pass, but there are far too many right now to actually run. Additionally, something goes wrong with the regexing for the metadata params, and I'm not sure why yet.

**Comment by Oddant1** [comment]
As of https://github.com/qiime2/q2-mystery-stew/pull/3/commits/1edb4bacac20e65daa8f1c1c564fc311f2a46fd5 some 127,944 usage examples are being registered. This is far too many to all be tested especially given the fact that the current implementation requires several accesses to the hard disk for each usage example, so I've just done some fiddling with smaller isolated test groups so far.

Based on this fiddling, I've determined that the metadata and collection parameters don't regex properly after being written to file even though the regex matches when I test it with external tools. Figuring out why this is happening is probably my next task. All other types appear to be working as intended though.

Additionally, there is currently no implementation for Tuples because they aren't Qiime2 primitives and thus can't be treated in quite the same way as all the other types I've implemented.  @ebolyen Just to make sure I'm not wasting time attempting to implement them I want to verify. We do want to test Tuples as an input type here correct? Currently the collections that are being tested are List and Set.

**Comment by thermokarst** [comment]
@Oddant1, just curious, are you able to run all 127,944 tests? If so, how long does it take on your dev machine?

**Comment by Oddant1** [comment]
@thermokarst I gave it about ten minutes then stopped it. I don't know how far it got.

**Comment by thermokarst** [comment]
Gotcha - thanks! Personally I'm okay with long tests here - this plugin is meant (I think) to be pretty exhaustive, I like to think of it as more of an "integration" suite, so probably not run with the same frequency as a typical QIIME 2 plugin.

File: q2_mystery_stew/plugin_setup.py [hunk 2]
```diff
@@ -18,3 +32,229 @@
...
+
+                    chosen_param_values.update({name: param_val})
+                else:
+                    domain_size = len(value.domain)
+                    chosen_param_values.update(
+                        {name: value.domain[sig.num_outputs % domain_size]})
+
+            usage_example = \
+                {action_name:
+                 UsageInstantiator(chosen_param_values, outputs, action_name)}
```
Comment (ebolyen) [inline q2_mystery_stew/plugin_setup.py:244]: I think we want one of these for each element in a type's domain. This gets wild with multiple inputs, so perhaps fix all other inputs at a single value while iterating through the domain of a given input, that way the number of usages matches the number of values if they were all independent.

**Comment by Oddant1** [comment]
Update on test run time performance. With the current implementation, I removed the collection and metadata involving tests which left me with 46002 usages. I gave these 48 minutes to run and they didn't finish.

File: q2_mystery_stew/plugin_setup.py (hunk 2 above)
Comment (Oddant1) [reply q2_mystery_stew/plugin_setup.py:244]: And do this for each combination of arguments we're generating? This caused me to have an idea. I'm not sure if this is what you meant or not, but what if we only had one function per number of outputs that took every parameter. Then we do what you said above. We create usage examples with every value in an individual argument's domain while holding all others in place. We do this for all arguments.

Additionally, under the current implementation we will end up with a usage example somewhere containing each value in a type's domain provided the domain is three or less elements (which they all are at the moment). It wouldn't be too hard to ensure a usage example if we were to end up with a parameter with a larger domain.

**Comment by thermokarst** [comment]
Thanks for the update, @Oddant1. Just for comparison sake, our current integration tests can take 120 to 150 minutes.

**Comment by Oddant1** [comment]
I discovered that I made some mistakes in previous commits that were causing the tests to behave in a different way than I expected. This commit cuts the number of test cases roughly in half, but now it should actually be doing what I thought it was before. This image shows the outputs of test cases generated when only the `single_int` param is used.
___
image
___

File: q2_mystery_stew/plugin_setup.py [hunk 3]
```diff
@@ -18,3 +32,229 @@
...
+                                           (-3.49, 0, 3.49))
+}
+float_values = tuple(float_params.values())
+
+collection_params = {
+    # collection parameters
+    'int_list': Param('int_list', List[Int], (int_values)),
+    'float_list': Param('float_list', List[Float], (float_values)),
+    'int_set': Param('int_set', Set[Int], (int_values)),
+    'float_set': Param('float_set', Set[Float], (float_values))
```
[the remark this answers is not in the document]
Comment (Oddant1) [reply q2_mystery_stew/plugin_setup.py:153]: Yeah, that's my end goal, but this implementation was much faster as a proof of concept.

File: q2_mystery_stew/plugin_setup.py [hunk 4]
```diff
@@ -18,3 +32,229 @@
...
+            if 'md' in name:
+                value = value.to_dataframe()
+            inputs_params = ''.join(f'{name}: {value}\n')
+
+        output1 = use.get_result('output_1')
+        output1.assert_has_line_matching(
+            label='validate inputs and params',
+            path='echo.txt',
+            expression=inputs_params,
+        )
```
[the remark this answers is not in the document]
Comment (Oddant1) [reply q2_mystery_stew/plugin_setup.py:68]: I'm currently smashing all the inputs and parameters together into one string then regexing the string that should contain all everything against the file.

**Comment by Oddant1** [comment]
All the fiddling with replacing braces with colons and stuff in https://github.com/qiime2/q2-mystery-stew/pull/3/commits/8885e9bf4c1362276633e23e0f847f55cfc95c77 is due to the fact that no matter how hard I tried I couldn't get the regex to match with the braces. I was trying to make it use a raw string to regex, and it just wan't having it. So I removed the braces.

**Comment by Oddant1** [comment]
https://github.com/qiime2/q2-mystery-stew/pull/3/commits/928d27e78bd838ce5c1a2fa8d7f3eaf034c1578d kinda hamfists an input into the tests just to demonstrate that it works. As suggested by @ebolyen the semantic types used as inputs in `q2-mystery-stew` will be (maybe slightly modified) reimplementations of the types currently being used for testing in the framework. This is to prevent coupling the mystery-stew tests to the framework.

**Comment by Oddant1** [comment]
https://github.com/qiime2/q2-mystery-stew/pull/3/commits/0df267669f587f80918d9c05a961e68622e4dbde starts adding rough implementation of types. There's some more that needs to be done for them to work fully. Additionally, I've thought of some ways I can probably clean up the params implementation a bit without changing the functionality.

**Comment by Oddant1** [comment]
As of https://github.com/qiime2/q2-mystery-stew/pull/3/commits/4c280f4d0981ce6825db1e3635024ea25d043543 I think we are at or nearing completion of the necessary test case params and inputs

**Comment by Oddant1** [comment]
I got a new laptop a few weeks ago, and today I decided I would start running these tests on it and just walk away and do something else. I have finally run all the usage examples created by the code in this PR and can actually confirm that they all work. It just took one hour forty minutes and eighteen seconds. Which could be a lot worse.
___
image
___

**Pull request merged**